src/aws-cpp-sdk-core/include/aws/core/client/AWSClientEventStreamingAsyncTask.h

```diff
@@ -0,0 +1,137 @@
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+#pragma once
+
+#include <aws/core/Core_EXPORTS.h>
+#include <aws/core/NoResult.h>
+#include <aws/core/client/AWSClient.h>
+#include <aws/core/utils/event/EventDecoderStream.h>
+#include <aws/core/utils/event/EventEncoderStream.h>
+#include <aws/core/utils/event/EventStreamHandler.h>
+#include <aws/core/utils/logging/ErrorMacros.h>
+
+namespace Aws {
+namespace Client {
+
+class AsyncCallerContext;
+
+/**
+ * A template class to create bi-directional eventstreaming async task.
+ * This is essentially a C++14 capture-by-move lambda
+ */
+template <typename OutcomeT, typename ClientT, typename AWSEndpointT, typename RequestT, typename HandlerT>
+class AWS_CORE_LOCAL BidirectionalEventStreamingTask final {
+ public:
+  BidirectionalEventStreamingTask(const ClientT* pClientThis, AWSEndpointT&& endpoint, const std::shared_ptr<RequestT>& pRequest,
+                                  const HandlerT& handler, const std::shared_ptr<const Aws::Client::AsyncCallerContext>& handlerContext,
+                                  const std::shared_ptr<Utils::Event::EventEncoderStream>& stream, const Http::HttpMethod method,
+                                  char const* const signerName)
+      : m_clientThis(pClientThis),
+        m_endpoint(std::move(endpoint)),
+        m_pRequest(pRequest),
+        m_handler(handler),
+        m_handlerContext(handlerContext),
+        m_stream(stream),
+        m_method(method),
+        m_signerName(signerName),
+        m_sem(Aws::MakeShared<Aws::Utils::Threading::Semaphore>("BidirectionalEventStreamingTask", 0, 1)) {
+    assert(m_clientThis);
+    assert(m_pRequest);
+    assert(m_signerName);
+
+    // Reset EventStream handler as it still has a raw dangling pointer of the original user request
+    m_pRequest->SetEventStreamHandler(m_pRequest->GetEventStreamHandler());
+
+    // Perform synchronization logic extracted from generated code
+    // Setup a signal to notify the submitter thread that the task has actually started
+    std::shared_ptr<Aws::Utils::Threading::Semaphore> sem = m_sem;
+    m_pRequest->SetRequestSignedHandler([stream, sem](const Aws::Http::HttpRequest& httpRequest) {
+      stream->SetSignatureSeed(Aws::Client::GetAuthorizationHeader(httpRequest));
+      sem->ReleaseAll();
+    });
+
+    std::weak_ptr<RequestT> wRequest = m_pRequest;
+    // Setup InitialResponse handler to use the new actual request object
+    if (!m_pRequest->GetHeadersReceivedEventHandler()) {
+      m_pRequest->SetHeadersReceivedEventHandler([wRequest](const Http::HttpRequest*, Http::HttpResponse* response) {
+        auto request = wRequest.lock();
+        if (!request) {
+          AWS_LOGSTREAM_FATAL(ClientT::GetAllocationTag(), "Unexpected nullptr bi-directional streaming request on initial response!");
+          assert(false);
+          return;
```

Comment:
nit i prefer no returns on a void function and to rely on if - else behavior i.e.
```
if (!request) {
  ....
} else if (!response) {
  ....
} else {
  .....
}
```